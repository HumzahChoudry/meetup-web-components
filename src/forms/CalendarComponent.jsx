/* eslint-disable no-return-assign */
import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import withErrorList from '../utils/components/withErrorList';

/**
 * @module CalendarComponent
 * inits flatpickr js date picker over a text input
*/
export class CalendarComponent extends React.Component {
	constructor(props) {
		super(props);

		this.onOpen = this.onOpen.bind(this);
		this.onClose = this.onClose.bind(this);

		this.onFlatpickrChange = this.onFlatpickrChange.bind(this);
	}

	/**
	* @description init the js date flatpickr component
	*/
	componentDidMount() {
		// flatpickr uses `window` on import,
		// which breaks on server-sider render.
		// lazy-loading flatpickr ensures it is
		// imported only in clientside envs.
		const Flatpickr = require('flatpickr');
		const options = {
			onChange: this.onFlatpickrChange,
			onOpen: this.onOpen,
			onClose: this.onClose,
			altInput: true,
			allowInput: true,
			altFormat: 'D M d, Y',
			defaultDate: this.props.value,
			nextArrow: `<span class="svg svg--chevron-right">
				<svg preserveAspectRatio="xMinYMin meet" width="12" height="12" viewBox="0 0 12 12" className="svg-icon valign--middle" role="img">
				<use xlink:href="#icon-chevron-right" />
				</svg>
			</span>`,
			prevArrow: `<span class="svg svg--chevron-left">
				<svg preserveAspectRatio="xMinYMin meet" width="12" height="12" viewBox="0 0 12 12" className="svg-icon valign--middle" role="img">
				<use xlink:href="#icon-chevron-left" />
				</svg>
			</span>`,
			...this.props.datepickerOptions,
		};

		this.flatpickr = new Flatpickr(this.inputEl, options);
	}

	componentWillUnmount() {
		this.flatpickr && this.flatpickr.destroy();
	}

	// replaces updateFlatpickr
	// if we receive a new value from parent, update Flatpickr
	componentWillReceiveProps(newProps) {
		if (this.flatpickr) {
			if (newProps.datepickerOptions) {
				this.updateFlatpickrOptions(newProps.datepickerOptions);
			}
			this.flatpickr.setDate(newProps.value);
		}
	}

	/**
	 * Updates config options on the flatpickr instance
	 * @param {Object} newOptions new configaration options for flatpickr
	 * @return {undefined}
	 */
	updateFlatpickrOptions(newOptions) {
		Object.keys(newOptions)
			.filter(
				option => this.props.datepickerOptions[option] !== newOptions[option]
			)
			.forEach(option => {
				this.flatpickr.set(option, newOptions[option]);
			});
	}

	/**
	* @function onFlatpickrChange
	* @param Array selectedDates
	* @param dateStr
	* @param Object instance the calendar instance
	* @description signature conforms to the onChange handler flatpickr expects
	* calls onChange if prop provided (eg from redux-form or wrapping components like DateTimePicker)
	*/
	onFlatpickrChange(selectedDates, dateStr, instance) {
		this.props.onChange && this.props.onChange(selectedDates[0]);
	}

	/**
	* @function onOpen
	* @description event hook for flatpickr, used to call onFocus
	* and apply focus highlight if this is a DateTimePicker
	*/
	onOpen() {
		this.props.onFocus && this.props.onFocus();
	}

	/**
	* @function onClose
	* @description event hook for flatpickr, used to call onBlur
	* and remove focus highlight if this is a DateTimePicker
	*/
	onClose() {
		this.props.onBlur && this.props.onBlur();
	}

	render() {
		const {
			className,
			name,
			label,
			required,
			value,
			datepickerOptions, // eslint-disable-line no-unused-vars
			onChange, // eslint-disable-line no-unused-vars
			helperText,
			...other
		} = this.props;

		const classNames = {
			label: cx(
				{
					required,
					'flush--bottom': helperText
				},
				className
			),
			helperText: cx(
				'helperTextContainer',
				{ required }
			),
			field: cx(
				'input--dateTimePicker select--reset',
				className
			)
		};


		return (
			<span>
				{label && (
					<label htmlFor={name} className={classNames.label}>
						{label}
					</label>
				)}
				{helperText &&
					<div className={classNames.helperText}>
						{helperText}
					</div>
				}
				<input
					type="text"
					id={name}
					name={name}
					defaultValue={value}
					className={classNames.field}
					ref={input => (this.inputEl = input)}
					{...other}
				/>
			</span>
		);
	}
}

CalendarComponent.propTypes = {
	name: PropTypes.string.isRequired,
	onChange: PropTypes.func, // provided by DateTimePicker or redux-form
	datepickerOptions: PropTypes.object,
	suppressError: PropTypes.bool,
	helperText: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	])
};

CalendarComponent.defaultProps = {
	datepickerOptions: {},
	suppressError: false,
};

export default withErrorList(CalendarComponent);
