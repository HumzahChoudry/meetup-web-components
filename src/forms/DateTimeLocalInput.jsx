import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

/**
 * @module DateTimeLocalInput
 * @description renders html5 datetime-local input
*/
class DateTimeLocalInput extends React.Component {

	constructor(props) {
		super(props);
		// this.state = {
		// 	value: this.props.value ? (new Date(this.props.value).toISOString()).split('.')[0] : ''
		// 	// example: 2017-02-18T00:00:00
		// 	// leaving off milliseconds
		// 	// datetime local wont set value with milliseconds
		// };
		this.onChange = this.onChange.bind(this);
	}

	/**
	* @function onChange
	* @description sets state and calls a callback with the value
	* (callback used in wrapping components like datetimepicker)
	* @param e Event the change event
	*/
	onChange(e) {
		this.props.onChange && this.props.onChange(e.target.value);
		this.props.onChangeCallback && this.props.onChangeCallback(e.target.value);
	}

	render() {
		const {
			id,
			label,
			className,
			required,
			value,
			error,
			onChangeCallback,	// eslint-disable-line no-unused-vars
			onChange, 			// eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'input--dateTimeLocal',
			className
		);

		const labelClassNames = cx({required});
		return (
			<div>
				<label htmlFor={id} className={labelClassNames}>{label}</label>
				<input
					id={id}
					type='datetime-local'
					value={value}
					className={classNames}
					onChange={this.onChange}
					required={required}
					{...other}
				/>
				{ error && <p className='text--error'>{error}</p> }
			</div>
		);

	}
}

DateTimeLocalInput.propTypes = {
	id: PropTypes.string,
	label: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element,
	]),
	onChangeCallback: PropTypes.func
};

export default DateTimeLocalInput;

