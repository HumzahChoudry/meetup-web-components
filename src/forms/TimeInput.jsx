import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

/**
* @module TimeInput
*/
class TimeInput extends React.Component {

	constructor(props) {
		super(props);

		this.onChange = this.onChange.bind(this);
	}

	/**
	* @function onChange
	* @param e Event Object
	* @description called when the input changes, in turn calls the onChange
	* 	handler prop, if there is one (eg supplied by redux-form) and an onChangeCallback
	*	which may be provided by a parent component such as DateTimePicker
	*/
	onChange(e) {
		this.props.onChange && this.props.onChange(e.target.value); // redux-form provides an onChange prop

		// onChangeCallback currently provided by a parent component eg DateTimePicker
		this.props.onChangeCallback && this.props.onChangeCallback(e.target.value);
	}

	render() {
		const {
			id,
			label,
			name,
			className,
			required,
			value,
			error,
			onChange,			// eslint-disable-line no-unused-vars
			onChangeCallback, 	// eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'input--time',
			className
		);

		const labelClassNames = cx(
			'label--field',
			{ required }
		);

		const errorId = `${id}-error`;

		if (error) {
			other['aria-invalid'] = true;
			other['aria-describedby'] = errorId;
		}

		return (
			<span>
				{ label && <label htmlFor={id} className={labelClassNames}>{label}</label> }
				<input
					id={id}
					type='time'
					name={name}
					value={value}
					className={classNames}
					required={required}
					onChange={this.onChange}
					ref={ input => this.inputEl = input }
					{...other}
				/>
				{ error && <p id={errorId} className='text--error'>{error}</p> }
			</span>
		);

	}
}

TimeInput.propTypes = {
	name: PropTypes.string.isRequired,
	error: PropTypes.string,
	label: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	]),
	required: PropTypes.bool,
	onChange: PropTypes.func,			// redux-form provides an onChange prop
	onChangeCallback: PropTypes.func, 	// currently provided by a parent component eg DateTimePicker
};

export default TimeInput;

