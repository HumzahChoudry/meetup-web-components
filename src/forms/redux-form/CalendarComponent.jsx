import React from 'react';
import CalendarComponent from '../CalendarComponent';

/**
 * @param {Object} props - props passed in from parent and redux-form
 * @description wraps standard TextInput web component for use with redux-form
 * deconstructs props that redux-forms sets and sets them on TextInput
 * @return {Component} TextInput
 */
const ReduxFormCalendarComponent = props => {
	const { meta, input, validateBeforeTouched, ...other } = props;
	const error = (validateBeforeTouched || meta.touched) ? meta.error : null;

	return <CalendarComponent error={error} {...input} {...other} />;
};

ReduxFormCalendarComponent.displayName = 'ReduxFormCalendarComponent';

export default ReduxFormCalendarComponent;
