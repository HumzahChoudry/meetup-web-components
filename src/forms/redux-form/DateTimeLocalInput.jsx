import React from 'react';
import DateTimeLocalInput from '../DateTimeLocalInput';

/**
 * This component wraps the standard DateTimeLocalInput
 * component for use with redux-form.
 * It deconstructs props that redux-form sets and re-sets them on DateTimeLocalInput
 *.
 * @param {object} props React component props
 * @return {React.Component} DateTimeLocalInput
 *
 */
const ReduxFormDateTimeLocalInput = props => {
	const { meta, input, ...other } = props;

	return <DateTimeLocalInput {...input} error={meta.error} {...other} />;
};

ReduxFormDateTimeLocalInput.displayName = 'ReduxFormDateTimeLocalInput';

export default ReduxFormDateTimeLocalInput;
