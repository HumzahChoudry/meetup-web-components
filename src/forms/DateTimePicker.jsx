import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import CalendarComponent from './CalendarComponent';
import TimeInput from './TimeInput';
import DateTimeLocalInput from './DateTimeLocalInput';

/**
 * @module DateTimePicker
 * @description a component that renders a calendar ui and time input
 * defaults to datetime-local input on supported mobile browsers.
 */
class DateTimePicker extends React.Component {

	constructor(props) {
		super(props);

		const datetime = props.value ? new Date(props.value) : new Date();

		this.state = {
			datetime,
			isDateTimeLocalSupported: false
		};

		this.getDate = this.getDate.bind(this);
		this.setDate = this.setDate.bind(this);

		this.getTime = this.getTime.bind(this);
		this.setTime = this.setTime.bind(this);

		this.setDateTime = this.setDateTime.bind(this);
		this.onFocus = this.onFocus.bind(this);
		this.onBlur = this.onBlur.bind(this);
	}
	/**
	/* call `hasBrowserSupport` after mounting so server
	/* and client-side renders match, then immediately
	/* force a re-render. otherwise, client-side render
	/* will default to the server render, which never has
	/* browser support.
	/* @returns {undefined}
	*/
	componentDidMount() {
		this.setState({ isDateTimeLocalSupported: this.hasBrowserSupport() });
	}

	/**
	* @function hasBrowserSupport
	* @description test if this browser supports datetime local
	* @return bool whether or not this browser supports datetime local
	*/
	hasBrowserSupport() {
		if (this.props.forceCalendar || typeof document === 'undefined') {
			return false;
		}

		const input = document.createElement('input'),
			invalidValue = 'notadate';

		// some browsers (as Android stock browsers) pretend they support
		// certain input types,
		// so we set the value to an invalid value and see if browser rejects

		input.setAttribute('type', 'datetime-local');
		input.setAttribute('value', invalidValue);

		return input.value !== invalidValue;
	}

	/**
	* @function getDate
	* @description gets a date from the datetime value in state
	* @return Date
	*/
	getDate() {
		if (!this.state.datetime) {
			return;
		}
		const datetime = new Date(this.state.datetime);
		return new Date(datetime.getFullYear(), datetime.getMonth(), datetime.getDate());
	}

	/**
	* @function setDate
	* @description sets the date portion of the state's datetime
	* @param Date value date to set in state
	*/
	setDate(value) {
		const newDate = new Date(value);
		const datetime = new Date(this.state.datetime);

		datetime.setFullYear(newDate.getFullYear(), newDate.getMonth(), newDate.getDate());
		this.setState({ datetime });

		this.dateComponent.setState({
			value: datetime
		});

		// update flatpickr on update
		this.dateComponent.updateFlatpickr();

		if (this.props.onChangeCallback) {
			this.props.onChangeCallback(datetime);
		}
	}

	/**
	* @function parseTimeFromDateTime
	* @description gets the time string part of our datetime state
	* @param Date datetime - datetime object to be parsed
	* @return String
	*/
	parseTimeFromDateTime(datetime) {
		return (datetime.toTimeString().split(' ')[0]).split(':', 2).join(':'); // TODO: localization/i18n SDS-247
	}

	/**
	* @function parseHoursAndMinutesFromTimeString
	* @description takes a time string and returns an array of hours and minutes
	* @param String timeString a 24-hour time string, eg `18:45`
	* @return Array [0] - the hours, [1] - the minutes
	*/
	parseHoursAndMinutesFromTimeString(timeString) {
		return timeString.split(':'); // TODO: localization/i18n SDS-247
	}


	/**
	* @function parseNewTimeAsDate
	* @description takes a time string in 24hr time (e.g. `18:35`)
	* and converts it to a Date object with the current state's date and timezone
	* @return Date new date with current date and the given time
	*/
	parseNewTimeAsDate(timeStr) {
		const { datetime } = this.state;
		const parsedTime = this.parseHoursAndMinutesFromTimeString(timeStr);

		datetime.setHours(parsedTime[0], parsedTime[1]);

		return datetime;
	}

	/**
	* @function getTime
	* @description gets the time portion of the state's datetime
	* @return String time portion of datetime  (hour and min)
	*/
	getTime() {
		const datetime = new Date(this.state.datetime);
		// split on space and leave off timezone and seconds
		return this.parseTimeFromDateTime(datetime);
	}

	/**
	* @function setTime
	* @description sets the time portion of the state's datetime
	* @param value String valid time string
	*/
	setTime(value) {
		const datetime = this.parseNewTimeAsDate(value);
		this.setDateTime(datetime);
	}

	/**
	* @function setDateTime
	* @description sets the datetime when changed
	* @param value  datetime
	*/
	setDateTime(value) {
		const datetime = new Date(value);

		this.setState({ datetime });

		if (this.props.onChangeCallback) {
			this.props.onChangeCallback(datetime);
		}
	}

	/**
	* @function onFocus
	* @param Event e
	* @description called if datetime compound component and used to set focused
	* class on combo component
	*/
	onFocus(e) {
		this.backgroundEl.classList.add('focused');
	}

	/**
	* @function onBlur
	* @param Event e
	* @description called when either of the date or time inputs is blurred
	* takes off focus class when neither of them are in focus
	*/
	onBlur(e) {
		const timeInput = this.timeComponent.inputEl;
		const dateInput = this.dateComponent.inputEl;
		if (document.activeElement !== timeInput && document.activeElement !== dateInput) {
			this.backgroundEl.classList.remove('focused');
		}
	}

	render() {
		const {
			className,
			dateOnly,
			datepickerOptions,	// eslint-disable-line no-unused-vars
			id,
			label,
			value,	// eslint-disable-line no-unused-vars
			required,
			name,
			error,
			...other
		} = this.props;

		const classNames = cx(
			'dateTimePicker',
			{calendarTimeComponent : !dateOnly },
			{required},
			{ 'field--error': error },
			className
		);
		const childClasses = cx(
			{ 'field--error': error }
		);
		const labelClassNames = cx({required});
		const timeInputName = `${name}-time`;

		if (this.state.isDateTimeLocalSupported) {
			// TODO datetime-local opts ?

			return (
				<div>
					<DateTimeLocalInput
						id={id}
						label={label}
						value={this.state.datetime}
						required={required}
						className={classNames}
						onChangeCallback={this.setDateTime}

						{...other} />
					{ error && <p className='text--error'>{error}</p> }
				</div>
			);
		}

		// only set callbacks if this is a combo datetime component
		const onFocus = (dateOnly) ? null : this.onFocus;
		const onBlur = (dateOnly) ? null : this.onBlur;

		return (
			<div>
				<label htmlFor={id} className={labelClassNames}>{label}</label>
				<div className={classNames}>

					<div>
						<CalendarComponent name={name}
							onChangeCallback={this.setDate}
							value={this.getDate()}
							onFocus={onFocus}
							onBlur={onBlur}
							opts={datepickerOptions}
							className={childClasses}
							ref={ comp => this.dateComponent = comp }
						/>
						{ !dateOnly &&
								<TimeInput name={timeInputName}
									onChange={this.setTime}
									onFocus={onFocus}
									onBlur={onBlur}
									value={this.getTime()}
									ref={ comp => this.timeComponent = comp }
									className={childClasses}
								/>
						}
						{ !dateOnly &&
							<input type='text'
								id='datetime-background'
								ref={ el => this.backgroundEl = el }
							/>
						}
						{ error && <p className='text--error'>{error}</p> }
					</div>
				</div>
			</div>
		);
	}
}

DateTimePicker.propTypes = {
	datepickerOptions: PropTypes.object,
	label: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.string
	]),
	name: PropTypes.string.isRequired,
	required: PropTypes.bool,
	value: PropTypes.oneOfType([
		PropTypes.number,
		PropTypes.object,
		PropTypes.string
	]),
	error: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	]),
	dateOnly: PropTypes.bool,
	forceCalendar: PropTypes.bool,
	onChangeCallback: PropTypes.func
};

export default DateTimePicker;
