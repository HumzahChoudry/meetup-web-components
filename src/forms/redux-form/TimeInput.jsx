import React from 'react';
import TimeInput from '../TimeInput';

/**
 * This component wraps the standard TimeInput for use with redux-form.
 * It deconstructs props that redux-form sets and re-sets them on TimeInput.
 * @param {object} props React component props
 * @return {React.Component} TimeInput
 */
const ReduxFormTimeInput = props => {
	const { input, meta, ...other } = props;
	console.log('HELLO HELLO HELLO');
	return <TimeInput {...input} error={meta.touched && meta.error} {...other} />;
};

ReduxFormTimeInput.displayName = 'ReduxFormTimeInput';

export default ReduxFormTimeInput;
