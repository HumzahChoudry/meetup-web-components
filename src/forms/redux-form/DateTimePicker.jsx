import React from 'react';
import DateTimePicker from '../DateTimePicker';

/**
 * @param {Object} props - props passed in from parent and redux-form
 * @description wraps standard TextInput web component for use with redux-form
 * deconstructs props that redux-forms sets and sets them on TextInput
 * @return {Component} TextInput
 */
const ReduxFormDateTimePicker = props => {
	const { meta, input, ...other } = props;
	let error = {};
	if (meta.touched) {
		error = { error: meta.error };
	}
	return <DateTimePicker {...error} {...input} {...other} />;
};

ReduxFormDateTimePicker.displayName = 'ReduxFormDateTimePicker';

export default ReduxFormDateTimePicker;
