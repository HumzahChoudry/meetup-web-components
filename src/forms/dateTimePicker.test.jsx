/* eslint-disable */
import React from 'react';
import { shallow, mount, render } from 'enzyme';
import DateTimePicker from './DateTimePicker';
import DateTimeLocalInput from './DateTimeLocalInput';
import CalendarComponent from './CalendarComponent';
import TimeInput from './TimeInput';

describe('DateTimePicker', function() {

	let dateTimeComponent;

	const 	year = 2018,
			month = 0,
			day = 1,
			hours = 23,
			minutes = 11;

	const datetime = new Date(Date.UTC(year, month, day, hours, minutes)),
			dateStr = `${year}-0${month+1}-0${day}`,
			minDate = new Date(),
			maxDate = new Date();
	
	maxDate.setDate(maxDate.getDate() + 5);

	const datepickerOptions = { minDate, maxDate };
	// timeProps

	describe('renders child components as expected', () => {
		beforeEach(() => {
			dateTimeComponent = shallow(
				<DateTimePicker
					id='event_start'
					name='start_time'
					value={datetime}
					error='Too late'
					datepickerOptions
					forceCalendar
				/>
			);
		});

		afterEach(() => {
			dateTimeComponent = null;
		});

		it('passes expected props and value to calendar and time', function() {
			// visually check:
			//  Calendar and TimeInput are in snapshot
			// name, id, required etc attrs areas expected
			// datepickerOpts in snapshot, with date range
			// datepickerCallback is passed down to each
			
			expect(dateTimeComponent).toMatchSnapshot();
		});

		it('sets datetime in state', function() {
			expect(dateTimeComponent.state('datetime')).toEqual(datetime);
		});

		it('has child components with prop.value equal to its state date and time', () => {
			dateTimeComponent = mount(
				<DateTimePicker 
					name='start_time'
					value={datetime}
					error='Too late'
					datepickerOptions
					forceCalendar
				/>
			);
			const timeComponent = dateTimeComponent.find(TimeInput);
			const calendarComponent = dateTimeComponent.find(CalendarComponent);
			const instance = dateTimeComponent.instance();
			
			expect(timeComponent.prop('value')).toEqual(instance.parseTimeFromDateTime(datetime));
			expect(calendarComponent.prop('value')).toEqual(instance.getDate());
		});


		it('renders a date component only if dateOnly specified, with no focus and blur props', function() {
			let dateComponentOnly = shallow(
				<DateTimePicker name='start_time'
					forceCalendar
					datepickerOptions
					dateOnly
				/>
			);
			expect(dateComponentOnly).toMatchSnapshot();
			dateComponentOnly = null;
		});

		it('renders a datetimelocal component if browser supports', function() {
			dateTimeComponent.instance().setState({ isDateTimeLocalSupported: true });
			const dateTimeLocalInput = dateTimeComponent.find(DateTimeLocalInput)
			expect(dateTimeLocalInput).not.toBeNull();
			// check value?

		});

		describe('accessibility concerns', () => {

			it('renders a fieldset for date and time, with optional offscreen legend for a11y', function() {
				let dateComponentOnly = shallow(
					<DateTimePicker name='start_time'
						forceCalendar
						datepickerOptions
						dateOnly
					/>
				);
				expect(dateComponentOnly).toMatchSnapshot();
				dateComponentOnly = null;
			});

			it('provides aria-invalid and aria-describedby attrs on calendar when there is an error', function() {
				let dateComponent = shallow(
					<DateTimePicker name='start_time'
						forceCalendar
						error='Too late'
					/>
				);
				expect(dateComponent).toMatchSnapshot();
			});
		})
	});



	describe('callbacks passed to child components get called', function() {
	// 	let focusSpy,
	// 		blurSpy,
	// 		dateInputEl,
	// 		timeInputEl;

	// 	beforeEach(() => {
	// 		focusSpy = spyOn(DateTimePicker.prototype, 'onFocus').and.callThrough();
	// 		blurSpy = spyOn(DateTimePicker.prototype, 'onBlur');
	// 		dateTimeComponent = TestUtils.renderIntoDocument(
	// 			<DateTimePicker name='start_time'
	// 				value={dateStr}
	// 				forceCalendar
	// 			/>
	// 		);
	// 		timeInputEl = TestUtils.findRenderedDOMComponentWithTag(dateTimeComponent.timeComponent, 'input');
	// 		dateInputEl = TestUtils.findRenderedDOMComponentWithTag(dateTimeComponent.dateComponent, 'input');
	// 	});

	// 	afterEach(() => {
	// 		focusSpy = null;
	// 		blurSpy = null;
	// 		dateInputEl = null;
	// 		timeInputEl = null;
	// 	});

	// 	it('has an onFocus callback that gets called when time input is focused', function() {
	// 		TestUtils.Simulate.focus(timeInputEl);
	// 		expect(focusSpy).toHaveBeenCalled();
	// 	});

	// 	it('has an onFocus callback that gets called when calendar input is focused', function() {
	// 		TestUtils.Simulate.focus(dateInputEl);
	// 		expect(focusSpy).toHaveBeenCalled();
	// 	});

	// 	it('has an onFocus callback that applied focused class', function() {
	// 		dateTimeComponent.onFocus();
	// 		expect(() => TestUtils.findRenderedDOMComponentWithClass(dateTimeComponent, 'focused')).not.toThrow();
	// 	});

	// 	it('has an onBlur callback that gets called when time input is blurred', function() {
	// 		TestUtils.Simulate.blur(dateInputEl);
	// 		expect(blurSpy).toHaveBeenCalled();
	// 	});

	// 	it('has an onBlur callback that gets called when calendar input is blurred', function() {
	// 		TestUtils.Simulate.blur(dateInputEl);
	// 		expect(blurSpy).toHaveBeenCalled();
	// 	});

	// });

	// describe('onChangeCallback\'s get called when child components change', function() {
	// 	it('has setTime callback called when time input changes', function() {
	// 		const setTimeSpy = spyOn(DateTimePicker.prototype, 'setTime');
	// 		dateTimeComponent = TestUtils.renderIntoDocument(
	// 			<DateTimePicker name='start_time'
	// 				value={dateStr}
	// 				forceCalendar
	// 			/>
	// 		);
	// 		const timeInputEl = TestUtils.findRenderedDOMComponentWithTag(dateTimeComponent.timeComponent, 'input');
	// 		TestUtils.Simulate.change(timeInputEl);
	// 		expect(setTimeSpy).toHaveBeenCalled();
	// 	});

	// 	it('has setDate callback called when date input changes', function() {
	// 		const setDateSpy = spyOn(DateTimePicker.prototype, 'setDate');
	// 		dateTimeComponent = TestUtils.renderIntoDocument(
	// 			<DateTimePicker name='start_time'
	// 				value={dateStr}
	// 				forceCalendar
	// 			/>
	// 		);

	// 		dateTimeComponent.dateComponent.flatpickr.setDate(dateStr, true);
	// 		expect(setDateSpy).toHaveBeenCalled();
	// 	});

	// 	it('has setDateTime callback called when datetime input changes', function() {
	// 		const setDateTimeSpy = spyOn(DateTimePicker.prototype, 'setDateTime');
	// 		dateTimeComponent = TestUtils.renderIntoDocument(
	// 			<DateTimePicker name='start_time'
	// 				value={dateStr}
	// 			/>
	// 		);
	// 		dateTimeComponent.setState({ isDateTimeLocalSupported: true }); // forcing datetimelocal
	// 		const dateTimeInputEl = TestUtils.findRenderedDOMComponentWithTag(dateTimeComponent, 'input');
	// 		TestUtils.Simulate.change(dateTimeInputEl);
	// 		expect(setDateTimeSpy).toHaveBeenCalled();
	// 	});

	// });

	// describe('onChangeCallback in props', () => {
	// 	const mockFn = jest.fn();

	// 	beforeEach(() => {
	// 		dateTimeComponent = TestUtils.renderIntoDocument(
	// 			<DateTimePicker name='start_time'
	// 				value={dateStr}
	// 				onChangeCallback={mockFn}
	// 				forceCalendar
	// 			/>
	// 		);
	// 	});

	// 	afterEach(() => {
	// 		mockFn.mockClear();
	// 	});

	// 	it('is called when time input changes', () => {
	// 		const dateTimeInputEl = TestUtils.findRenderedDOMComponentWithTag(dateTimeComponent.timeComponent, 'input');
	// 		TestUtils.Simulate.change(dateTimeInputEl);
	// 		expect(mockFn).toHaveBeenCalledWith(expect.any(Date));
	// 	});

	// 	it('is called when date input changes', () => {
	// 		dateTimeComponent.dateComponent.flatpickr.setDate(dateStr, true);
	// 		expect(mockFn).toHaveBeenCalledWith(expect.any(Date));
	// 	});
	});

});
	



