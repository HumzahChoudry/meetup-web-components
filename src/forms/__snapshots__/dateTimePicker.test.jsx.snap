// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker renders child components as expected accessibility concerns provides aria-invalid and aria-describedby attrs on calendar when there is an error 1`] = `
<div>
  <fieldset>
    <Flex
      direction="row"
    >
      <FlexItem>
        <CalendarComponent
          aria-describedby="calendar-error-1503418845095"
          aria-invalid={true}
          className="field--error"
          datetimePickerCallback={[Function]}
          name="start_time"
          onBlur={[Function]}
          onFocus={[Function]}
          value={2017-08-22T04:00:00.000Z}
        />
      </FlexItem>
      <FlexItem>
        <TimeInput
          className="field--error"
          datetimePickerCallback={[Function]}
          id="undefined-time"
          name="start_time-time"
          onBlur={[Function]}
          onFocus={[Function]}
          value="12:20"
        />
      </FlexItem>
    </Flex>
  </fieldset>
  <p
    className="text--error"
    id="calendar-error-1503418845095"
  >
    Too late
  </p>
</div>
`;

exports[`DateTimePicker renders child components as expected accessibility concerns renders a fieldset for date and time, with optional offscreen legend for a11y 1`] = `
<div>
  <CalendarComponent
    className=""
    datetimePickerCallback={[Function]}
    name="start_time"
    onBlur={null}
    onFocus={null}
    opts={true}
    value={2017-08-22T04:00:00.000Z}
  />
</div>
`;

exports[`DateTimePicker renders child components as expected passes expected props and value to calendar and time 1`] = `
<div>
  <fieldset>
    <Flex
      direction="row"
    >
      <FlexItem>
        <CalendarComponent
          aria-describedby="calendar-error-1503418844941"
          aria-invalid={true}
          className="field--error"
          datetimePickerCallback={[Function]}
          id="event_start"
          name="start_time"
          onBlur={[Function]}
          onFocus={[Function]}
          opts={true}
          value={2018-01-01T05:00:00.000Z}
        />
      </FlexItem>
      <FlexItem>
        <TimeInput
          className="field--error"
          datetimePickerCallback={[Function]}
          id="event_start-time"
          name="start_time-time"
          onBlur={[Function]}
          onFocus={[Function]}
          value="18:11"
        />
      </FlexItem>
    </Flex>
  </fieldset>
  <p
    className="text--error"
    id="calendar-error-1503418844941"
  >
    Too late
  </p>
</div>
`;

exports[`DateTimePicker renders child components as expected renders a date component only if dateOnly specified, with no focus and blur props 1`] = `
<div>
  <CalendarComponent
    className=""
    datetimePickerCallback={[Function]}
    name="start_time"
    onBlur={null}
    onFocus={null}
    opts={true}
    value={2017-08-22T04:00:00.000Z}
  />
</div>
`;
